Feat: Gateway Upgrade Refactoring + LTS upgrade guide

lena-larionova opened the pull request:
### Description

This PR add general Gateway upgrade docs and an LTS upgrade guide. These two pieces of content are extremely interdependent, so they have to be published together.

Diagrams: I'll probably publish the initial MVP guides using the diagrams here, which have been provided by eng. Will remake as a fast-follow for style.

https://konghq.atlassian.net/browse/DOCU-3244
https://konghq.atlassian.net/browse/DOCU-3193

### Testing instructions

https://deploy-preview-6534--kongdocs.netlify.app/gateway/latest/upgrade/lts-upgrade/
https://deploy-preview-6534--kongdocs.netlify.app/gateway/latest/upgrade/
https://deploy-preview-6534--kongdocs.netlify.app/gateway/latest/upgrade/dual-cluster/
https://deploy-preview-6534--kongdocs.netlify.app/gateway/latest/upgrade/in-place/
https://deploy-preview-6534--kongdocs.netlify.app/gateway/latest/upgrade/blue-green/
https://deploy-preview-6534--kongdocs.netlify.app/gateway/latest/upgrade/rolling-upgrade/
https://deploy-preview-6534--kongdocs.netlify.app/gateway/latest/upgrade/backup-and-restore/

### Checklist 

- [x] Review label added <!-- (see below) -->
- [x] PR pointed to correct branch (`main` for immediate publishing, or a release branch: e.g. `release/gateway-3.2`, `release/deck-1.17`)


<!-- !!! Only Kong employees can add labels due to a GitHub limitation. If you're an OSS contributor, thank you! The maintainers will label this PR for you !!! -->

<!-- When raising a pull request, indicate what type of review you need with one of the following labels:

    review:copyedit: Request for writer review.
    review:general: Review for general accuracy and presentation. Does the doc work? Does it output correctly?
    review:tech: Request for technical review for a docs platform change.
    review:sme: Request for review from an SME (engineer, PM, etc).

At least one of these labels must be applied to a PR or the build will fail.
-->

outsinre commented:
@lena-larionova just add a reminder.

When drafting this PR, we'd better keep in mind the following two scenarios:

1. A general guide for upgrade between any Gateway versions.
2. A specific guide for upgrade between LTS `2.8` and LTS `3.4`.

lena-larionova commented:
> When drafting this PR, we'd better keep in mind the following two scenarios:
> 
> 1. A general guide for upgrade between any Gateway versions.
> 2. A specific guide with specific details for upgrade between LTS `2.8` and LTS `3.4`.

Yes, that's why in the draft, I have it as two guides:
https://deploy-preview-6534--kongdocs.netlify.app/gateway/latest/upgrade/lts-upgrade/
https://deploy-preview-6534--kongdocs.netlify.app/gateway/latest/upgrade/

cloudjumpercat commented on app/_data/docs_nav_gateway_3.4.x.yml:
```diff
@@ -29,8 +29,10 @@ items:
       - text: Breaking Changes	
         items:
-          - text: Kong Gateway 3.4.x	
+          - text: List of versions
```
Is this page meant to be as a "redirect"/placeholder since the old pages used to point here? Since we already list the breaking changes by version below, I don't think we need this in the nav. Could we keep the page so people who follow that old link end up here but the page is unlisted in the nav?

cloudjumpercat replied:
From Lena:

> how about turning the breaking changes section into a "book" so that it gets next/previous links? And then maybe use them in variables at the top of each breaking changes page as well.

I like this idea, I think it would work!

cloudjumpercat commented on app/_data/docs_nav_gateway_3.4.x.yml:
```diff
@@ -230,11 +232,26 @@ items:
+                url: /upgrade/blue-green/
+              - text: Rolling Upgrade
+                url: /upgrade/rolling-upgrade/
+          - text: Backup and Restore
```
I think it might make sense to move this page before the Upgrade Strategies since people would want to do this before they follow the upgrade steps.

cloudjumpercat commented on app/_data/docs_nav_gateway_3.5.x.yml:
```diff
@@ -234,11 +235,26 @@ items:
+                url: /upgrade/blue-green/
+              - text: Rolling Upgrade
+                url: /upgrade/rolling-upgrade/
+          - text: Backup and Restore
```
I think it might make sense to move this page before the Upgrade Strategies since people would want to do this before they follow the upgrade steps.

cloudjumpercat commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,148 @@
+
+Before upgrading, you should back up data with database tools and optionally declarative tools. We highly recommend doing backup in both methods, which offers you recovery flexibility. The database tools are robust and can restore data instantly compared to the declarative tools. Consequently, in case of data corruption, try to do a database-level restore first, and otherwise bootstrap a new data ...
+
+## Database
```
Probably a small suggestion, but I think maybe separating the sections by backup and restore vs. database and declarative would work?

## Backup
### database
### declarative
## Restore
### database
### declarative

Just because everyone needs to backup, they might not need to restore, so best to put that at the bottom and separate from backup in case they don't need it?

cloudjumpercat commented on app/_data/docs_nav_gateway_3.4.x.yml:
```diff
@@ -230,11 +232,26 @@ items:
+                url: /upgrade/rolling-upgrade/
+          - text: Backup and Restore
+            url: /upgrade/backup-and-restore/
+          - text: Upgrade Considerations
```
I think I'd also bump this one up to before the Upgrade Strategies section

cloudjumpercat commented on app/_includes/lts-changes.html:
```diff
@@ -0,0 +1,28 @@
+  <thead>
+    <th>Change</th>
+    <th>Category</th>
+    <th>Action required</th>
```
Is it possible to remove the ones that do not require an action from the table? Just thinking that if people don't need to perform an action, they probably don't care what the new feature is?

lena-larionova replied:
Unfortunately no, but I can remove the column. This is one of the categories of items that, for the LTS upgrade, they wanted to be sure users are notified - the idea is that they shouldn't need to look through the changelog and parse what's important for the LTS, since it's all covered here.

cloudjumpercat commented on app/_src/gateway/upgrade/lts-upgrade.md:
```diff
@@ -0,0 +1,217 @@
+2. Next, upgrade DP nodes using the recommendations from the section [DB-less Mode](#db-less-mode). 
+Point the new DP nodes to the new CP to avoid version conflicts.
+
+## Preparation: Review gateway changes
```
I wonder if it would be better to move these to the breaking changes section as a 2.8 LTS to 3.4 LTS breaking changes page of its own. That would be consistent with the other pages.

cloudjumpercat replied:
For the "Notifications of changed behavior" subsection, can we just tell them to read the release notes especially if this is just for awareness and not for actionable steps? Just trying to avoid duplicate content/too much content on one page.

lena-larionova replied:
> I wonder if it would be better to move these to the breaking changes section as a 2.8 LTS to 3.4 LTS breaking changes page of its own. That would be consistent with the other pages.

I would be fine with this, not sure how it would go over though. Let's try it out.

> For the "Notifications of changed behavior" subsection, can we just tell them to read the release notes especially if this is just for awareness and not for actionable steps? Just trying to avoid duplicate content/too much content on one page.

Unfortunately no, for the same reason that I put for your previous comment - the goal of the LTS guide is to consolidate all the info they might need, and reading through the changelog is far more difficult.

cloudjumpercat reviewed (commented):
Left some structural feedback on topics, let me know if you have any questions!

cloudjumpercat commented on app/_src/gateway/upgrade/lts-upgrade.md:
```diff
@@ -0,0 +1,217 @@
+
+**Preparation phase**
+
+There are a number of steps you must complete before upgrading to Kong 3.4.x.x LTS:
```
```suggestion
There are a number of steps you must complete before upgrading to {{site.base_gateway}} 3.4.x.x LTS:
```

cloudjumpercat commented on app/_src/gateway/upgrade/lts-upgrade.md:
```diff
@@ -0,0 +1,217 @@
+1. Work through any listed prerequistes.
+1. Back up your database (if applicable), or your declarative configuration files.
+1. Choose the right strategy for upgrading based on your deployment topology.
+1. Review the gateway changes from 2.8 to 3.4 for any breaking changes that affect your deployments.
```
```suggestion
1. Review the {{site.base_gateway}} changes from 2.8 to 3.4 for any breaking changes that affect your deployments.
```

cloudjumpercat commented on app/_src/gateway/upgrade/lts-upgrade.md:
```diff
@@ -0,0 +1,217 @@
+
+**Performing the upgrade**
+
+The actual execution of the migration is dependent on the type of deployment you have with Kong. 
```
```suggestion
The actual execution of the migration is dependent on the type of deployment you have with {{site.base_gateway}}. 
```

cloudjumpercat commented on app/_src/gateway/upgrade/lts-upgrade.md:
```diff
@@ -0,0 +1,217 @@
+A traditional mode deployment is when all {{site.base_gateway}} components are running in one environment, and there is no control plane/data plane separation.
+
+You have two options when upgrading {{site.base_gateway}} in traditional mode:
+* **Dual-cluster upgrade**: A new Kong cluster of version Y is deployed alongside the current version X, so that two clusters serve requests concurrently during the upgrade process.
```
```suggestion
* **Dual-cluster upgrade**: A new {{site.base_gateway}} cluster of version Y is deployed alongside the current version X, so that two clusters serve requests concurrently during the upgrade process.
```

cloudjumpercat commented on app/_src/gateway/upgrade/lts-upgrade.md:
```diff
@@ -0,0 +1,217 @@
+
+#### Dual-cluster upgrade
+
+Upgrading a Kong API Gateway from one LTS version to another LTS version with zero downtime can be achieved through a dual-cluster deployment strategy. This approach involves setting up a new cluster running the upgraded version of Kong alongside the existing cluster running the current version.
```
```suggestion
Upgrading a Kong API Gateway from one LTS version to another LTS version with zero downtime can be achieved through a dual-cluster deployment strategy. This approach involves setting up a new cluster running the upgraded version of {{site.base_gateway}} alongside the existing cluster running the current version.
```

cloudjumpercat commented on app/_src/gateway/upgrade/lts-upgrade.md:
```diff
@@ -0,0 +1,217 @@
+
+### DB-less mode
+
+DB-less mode is special in that each independent Kong node loads a copy of Kong configuration data into memory, without persistent database storage, so failure of some nodes will not spread to others.
```
```suggestion
DB-less mode is special in that each independent Kong node loads a copy of Kong configuration data into memory, without persistent database storage, so if some nodes fail it will not spread to others.
```

cloudjumpercat commented on app/_src/gateway/upgrade/rolling-upgrade.md:
```diff
@@ -0,0 +1,71 @@
+
+![Rolling upgrade workflow](/assets/images/products/gateway/upgrade/rolling-upgrade.png)
+
+> _Figure 1: Rolling upgrade workflow_
```
Can we add more description here?

cloudjumpercat commented on app/_src/gateway/upgrade/rolling-upgrade.md:
```diff
@@ -0,0 +1,71 @@
+
+## Prerequisites
+
+* You have reviewed the [general upgrade guide](/gateway/{{page.kong_version}}/upgrade/).
```
```suggestion
* Review the [general upgrade guide](/gateway/{{page.kong_version}}/upgrade/).
```

cloudjumpercat commented on app/_src/gateway/upgrade/rolling-upgrade.md:
```diff
@@ -0,0 +1,71 @@
+## Prerequisites
+
+* You have reviewed the [general upgrade guide](/gateway/{{page.kong_version}}/upgrade/).
+* You have chosen this upgrade option because you have a DB-less deployment, or you need to 
```
```suggestion
* You have a DB-less deployment or you need to 
```

cloudjumpercat commented on app/_src/gateway/upgrade/rolling-upgrade.md:
```diff
@@ -0,0 +1,71 @@
+
+The following steps follow the process outlined in the diagram, using DB-less mode as the example.
+
+1. Stop any {{site.base_gateway}} configuration updates (e.g. Admin API calls). 
```
```suggestion
1. Stop any {{site.base_gateway}} configuration updates ( Admin API calls for example). 
```

cloudjumpercat commented on app/_src/gateway/upgrade/rolling-upgrade.md:
```diff
@@ -0,0 +1,71 @@
+2. Back up data from the current cluster Y by following the 
+[declarative configuration backup instructions](/gateway/{{page.kong_version}}/upgrade/backup-and-restore/#declarative-config-backup).
+
+3. Evaluate factors that may impact the upgrade, as described in [Upgrade Considerations].
```
missing link

cloudjumpercat commented on app/assets/images/products/gateway/upgrade/blue-green-upgrade.png:
I'm not sure what the "admin writes" means, or why "kong migrations finish" is strikethrough

cloudjumpercat commented on app/assets/images/products/gateway/upgrade/hybrid-mode-upgrade.png:
This diagram and others, what does the dotted line mean vs the regular line?

outsinre replied:
Seems the diagram is removed. It is split into 4 separate graphs.

cloudjumpercat reviewed (changes requested):
Left some line edits! Let me know if you have any questions! I think there were also a few places that had "Kong" or "Gateway" too that I didn't mark since it's probably easier for you to just do a find and replace with {{site.base_gateway}}.

lena-larionova commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,153 @@
+
+    Refer to the [`kong migrations` CLI reference](/gateway/{{page.kong_version}}/reference/cli/#kong-migrations) for more information.
+
+4. Corresponding to the Postgres database dump, you can now restore the data with the utility `pg_restore`:
```
I think that whole first section before the comma can go, it doesn't provide any value.

lena-larionova commented on app/_src/gateway/upgrade/dual-cluster.md:
```diff
@@ -0,0 +1,92 @@
+![Dual-cluster upgrade workflow](/assets/images/products/gateway/upgrade/dual-cluster-upgrade.png)
+> _Figure 1: Dual-cluster upgrade workflow_
+
+This method has limitations on automatically generated runtime metrics that rely on the database. 
```
Honestly, I'm not really sure either. It comes from the upgrade guide and I don't really understand what it refers to. 
@outsinre could you please clarify?

outsinre replied:
@lena-larionova @cloudjumpercat 

Since two clusters with two databases. So some runtime metrics (like number of requests) are sent to two databases separately. Therefore, the metrics between the databases are not synced. If there is only one database, this is not an issue.

outsinre commented:
Let me do it this week!

outsinre commented:
in progress ...

outsinre commented on app/_src/gateway/upgrade/index.md:
```diff
@@ -0,0 +1,400 @@
+
+The following diagrams, figure 4 and 5, are the counterparts of figure 2 and 3 respectively. 
+
+Using the dual-cluster strategy with a rolling upgrade workflow:
```
Add link to "the dual-cluster strategy"

outsinre commented on app/_src/gateway/upgrade/index.md:
```diff
@@ -0,0 +1,400 @@
+> _Figure 4: Upgrading using the dual-cluster and rolling strategies. The diagram shows the new CP Y, deployed alongside with current CP X, while current DP nodes X are still serving API requests._
+_In the image, the background color of the current CP X or current DB is white instead of blue, signaling that the CP part was already upgraded and might have been decommissioned._
+
+Using the in-place cluster strategy with a rolling upgrade workflow:
```
Add a link to "in-place strategy".

```suggestion
Using the in-place strategy with a rolling upgrade workflow:
```

outsinre commented on app/_src/gateway/upgrade/index.md:
```diff
@@ -0,0 +1,327 @@
+
+Using the in-place cluster strategy with a rolling upgrade workflow:
+![In-place and rolling upgrade workflow](/assets/images/products/gateway/upgrade/in-place-rolling-hybrid-upgrade.png)
+> _Figure 5: Upgrade by in-place strategy and rolling strategy_
```
We can add a sentence like:

```
For details regarding each upgrade strategy, please refer to ...
```

outsinre commented on app/_src/gateway/upgrade/index.md:
```diff
@@ -0,0 +1,400 @@
+![In-place and rolling upgrade workflow](/assets/images/products/gateway/upgrade/in-place-rolling-hybrid-upgrade.png)
+> _Figure 5: Upgrade by in-place strategy and rolling strategy_
+
+## Prepation: Review breaking changes
```
```suggestion
## Preparation: Review breaking changes
```

outsinre commented on app/_src/gateway/upgrade/index.md:
```diff
@@ -0,0 +1,400 @@
+Selecting a strategy for the target deployment mode doesn't guarantee that you can start the upgrade immediately.
+You must also account for the following factors:
+
+* During the upgrade process, no changes can be made to the configuration database. 
```
```suggestion
* During the upgrade process, no changes can be made to the database. 
```

outsinre commented on app/_src/gateway/upgrade/index.md:
```diff
@@ -0,0 +1,400 @@
+* If you have custom plugins, review the code against changelog and test the custom plugin using the new version Y.
+* If you have modified any Nginx templates like `nginx-kong.conf` and `nginx-kong-stream.conf`, also make those changes to the templates for the new version Y. 
+Refer to [Nginx Directives](/gateway/{{page.kong_version}}/reference/nginx-directives/) for a detailed customization guide.
+* If you're using {{site.ee_product_name}}, make sure to [apply the enterprise license] to the new Gateway cluster.
```
"apply the enterprise license" misses the link.

outsinre commented on app/_src/gateway/upgrade/index.md:
```diff
@@ -0,0 +1,400 @@
+* If you have modified any Nginx templates like `nginx-kong.conf` and `nginx-kong-stream.conf`, also make those changes to the templates for the new version Y. 
+Refer to [Nginx Directives](/gateway/{{page.kong_version}}/reference/nginx-directives/) for a detailed customization guide.
+* If you're using {{site.ee_product_name}}, make sure to [apply the enterprise license] to the new Gateway cluster.
+* Always remember take a backup.
```
add a link to "backup"

outsinre commented on app/_src/gateway/upgrade/dual-cluster.md:
```diff
@@ -0,0 +1,96 @@
+
+With a dual-cluster upgrade, you deploy a new cluster of version Y alongside the current version X, 
+so that two clusters serve requests concurrently during the upgrade process. 
+As you add nodes, you will gradually adjust the traffic ratio between the two clusters to 
```
"As you add Kong Gatewa nodes" be better?

outsinre commented on app/_src/gateway/upgrade/dual-cluster.md:
```diff
@@ -0,0 +1,96 @@
+> _Figure 1: Dual-cluster upgrade workflow_
+
+This method has limitations on automatically generated runtime metrics that rely on the database. 
+For example, if the Rate-Limiting-Advanced (RLA) plugin is configured to store request counters in 
+the database, the counters between database X and database Y are not synchronized. 
+The impact scope depends on the `window_size` parameter of the plugin. 
```
Another example:

Similarly, the same limitation applies to [Vitals](https://docs.konghq.com/gateway/latest/kong-enterprise/analytics/) if you have a large amount of buffered metrics in Postgres or Cassandra.

lena-larionova replied:
This is in here, it's just version-gated. But in hindsight, since this is an upgrade doc, I think it's fair to leave the example unversioned, since they could be coming from any version, where Vitals and Cassandra are both still supported.

outsinre commented on app/_src/gateway/upgrade/dual-cluster.md:
```diff
@@ -0,0 +1,96 @@
+1. Stop any {{site.base_gateway}} configuration updates (e.g. Admin API calls). 
+This is critical to guarantee data consistency between cluster X and cluster Y.
+
+    To keep data consistency between the two clusters, you must not execute any write operations through Admin API, Kong Manager, or direct database updates. 
```
```suggestion
    To keep data consistency between the two clusters, you must not execute any write operations through Admin API, decK, Kong Manager, or direct database updates. 
```

outsinre commented on app/_src/gateway/upgrade/dual-cluster.md:
```diff
@@ -0,0 +1,96 @@
+3. Evaluate factors that may impact the upgrade, as described in [Upgrade considerations](/gateway/{{page.kong_version}}/upgrade/#preparation-upgrade-considerations/).
+You may have to consider customization of both `kong.conf` and {{site.base_gateway}} configuration data.
+
+4. Evaluate any [breaking changes](/gateway/{{page.kong_version}}/breaking-changes/) that may 
```
Not only breaking changes, but also the complete changelog: https://docs.konghq.com/gateway/changelog/

outsinre commented on app/_src/gateway/upgrade/dual-cluster.md:
```diff
@@ -0,0 +1,96 @@
+    3. [Restore the backup data](/gateway/{{page.kong_version}}/upgrade/backup-and-restore/#restore)
+    to the new database.
+
+    4. Configure the new cluster Y to point to the new database.
```
Replace this No.4 with No.1, and remove No.1. So No.2 will be the first step.

Important!

outsinre commented:
Still in progress, will add more comments

outsinre commented on app/_src/gateway/upgrade/dual-cluster.md:
```diff
@@ -0,0 +1,96 @@
+
+5. Deploy a new {{site.base_gateway}} cluster of version Y:
+
+    1. Install a new {{site.base_gateway}} cluster running version Y as instructed in the 
+    [{{site.base_gateway}} Installation Options](/gateway/{{page.kong_version}}/install/) and 
+    point it at the existing database for cluster X.
```
```suggestion
    1. Install a new {{site.base_gateway}} cluster running version Y as instructed in the 
    [{{site.base_gateway}} Installation Options](/gateway/{{page.kong_version}}/install/).
```

This step only install Kong Gateway, not start yet.

outsinre commented on app/_src/gateway/upgrade/rolling-upgrade.md:
```diff
@@ -0,0 +1,72 @@
+
+## Upgrade using the rolling method
+
+This guide refers to the old version as cluster X and the new version as cluster Y.
```
Move to the beginning of this page?

outsinre commented on app/_src/gateway/upgrade/rolling-upgrade.md:
```diff
@@ -0,0 +1,72 @@
+3. Evaluate factors that may impact the upgrade, as described in [Upgrade considerations](/gateway/{{page.kong_version}}/upgrade/#preparation-upgrade-considerations/).
+You may have to consider customization of both `kong.conf` and {{site.base_gateway}} configuration data.
+
+4. Evaluate any [breaking changes](/gateway/{{page.kong_version}}/breaking-changes/) that may 
+have happened between releases.
```
Add https://docs.konghq.com/gateway/changelog/ as well?

outsinre commented on app/_src/gateway/upgrade/rolling-upgrade.md:
```diff
@@ -0,0 +1,72 @@
+    [{{site.base_gateway}} Installation Options](/gateway/{{page.kong_version}}/install/) and 
+    point it at the existing database for cluster X.
+    
+    2. Perform staging tests against version Y to make sure it works for all use cases. 
+    
+    For example, does the Key Authentication plugin authenticate requests properly?
+
+    If it is a data plane node, ensure the communication with the control node succeeds.
+
+    If the outcome is not as expected, look over the 
+    [upgrade considerations](/gateway/{{page.kong_version}}/upgrade/#preparation-upgrade-considerations/) and the 
+    [breaking changes](/gateway/{{page.kong_version}}/breaking-changes/)
+    again to see if you missed anything.
+
+    3. Continuously install and launch more Y nodes.
```
Format error here.

outsinre commented on app/_src/gateway/upgrade/rolling-upgrade.md:
```diff
@@ -0,0 +1,72 @@
+
+5.  Deploy a new {{site.base_gateway}} cluster of version Y:
+    
+    1. Install a new cluster running version Y as instructed in the 
+    [{{site.base_gateway}} Installation Options](/gateway/{{page.kong_version}}/install/) and 
+    point it at the existing database for cluster X.
```
```suggestion
    1. Install a new cluster running version Y as instructed in the 
    [{{site.base_gateway}} Installation Options](/gateway/{{page.kong_version}}/install/).
```

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,157 @@
+
+We recommend backing up data using both methods, as this offers recovery flexibility.
+* The native database tools are robust and can restore data instantly, compared to the declarative tool. 
+* In case of data corruption, try to do a database-level restore first, otherwise bootstrap a new database and use decK to load in entity data.
```
Apart from decK, we have `kong config` command as well. Seems the `kong config` CLI part is removed?

lena-larionova replied:
Revisiting this. Previous guidance from Rob (on many tickets that I won't track down now) was that we should not document `kong config` as a viable option _at all_, but based on Rob's response [here](https://docs.google.com/document/d/1ddblT7Nb34LfjGLSoIxKQuN0Q87TKAr8dZffoaID2zw/edit?disco=AAAA3sXNSIY), sounds like we can consider it as an okay secondary option.

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,157 @@
+* **Entities managed by decK**: decK does not manage Enterprise-only entities, like RBAC roles, credentials, keyring, licence, etc. Configure these security related entities separately using Admin API or Kong Manager.
+See the reference for [Entities managed by decK](/deck/latest/reference/entities/) for a full list.
+
+Due to these limitations, we recommend always using the [database dump method](#database-backup) in deployments using a database.
```
```suggestion
Due to these limitations, we recommend prioritize the [database native method](#database-backup) in deployments using a database.
```

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
````diff
@@ -0,0 +1,157 @@
+    pg_database = kong
+    ```
+
+3. Bootstrap database entities by running `kong migrations bootstrap`.
+
+    Refer to the [`kong migrations` CLI reference](/gateway/{{page.kong_version}}/reference/cli/#kong-migrations) for more information.
````
Should we suppress the empty line for the two sentences to be in 1 line?

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,157 @@
+
+### Declarative restore
+
+If you need to roll back, change the DB-less {{site.base_gateway}} instance back to the original version, 
+validate the declarative config, then sync it back with your {{site.base_gateway}} instance.
```
decK is only for DB-mode.

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,157 @@
+
+For technical details, refer to the [manual backup method](/gateway/latest/kong-enterprise/db-encryption/#manual-export-and-manual-import) and the [automatic backup method](/gateway/latest/kong-enterprise/db-encryption/#automatic-backup-and-manual-recovery).
+
+## kong.conf backup
```
This section contains backup of files other than `kong.conf`. Should we change the title?

lena-larionova commented on app/_src/gateway/upgrade/index.md:
```diff
@@ -0,0 +1,400 @@
+* **Database backup** (_Traditional mode and control planes in hybrid mode_): 
+PostgreSQL has native exporting and importing tools that are reliable and performant, and that ensure 
+consistency when backing up or restoring data.
+* **Declarative backup** (_DB-less mode and data planes in hybrid mode_): In DB-less mode, configuration 
+is managed declaratively using a tool called decK. decK allows you to import and export configuration 
+using YAML or JSON files. 
```
This may be true, but we can't document `db_export` or `db_import` as viable strategies, as they're not recommended to use for backups.

lena-larionova commented:
@outsinre 

> We can add reference links to related pages like decK, Konnect, etc.

Done

> A bit different than the general guide in multiple places like https://github.com/Kong/docs.konghq.com/pull/6534#discussion_r1416675555, https://github.com/Kong/docs.konghq.com/pull/6534#discussion_r1416733336, etc.

Added a bunch of includes to reuse content so that it's the same across both the general and the LTS upgrade guides.

> May need to adjust the format like https://github.com/Kong/docs.konghq.com/pull/6534#discussion_r1416740350.
> Have empty lines that can be suppressed in multiple places. The format itself is not a syntax issue, but may change the upgrade steps.

Done 

> The new style of graphs looks much better.

Thanks :) I also went through and turned all of the remaining upgrade diagrams into the same style.


I've also applied the rest of your feedback, and added in the Kong CLI instructions + reworked the backup guide. I definitely interpreted the info wrong the first time, with too many inputs from previous support cases. Should be good now. Please take another look - hope we're close here 🤞 


@cloudjumpercat I've applied all your feedback as well, and reworked the diagrams with better captions. Please take another look when you can.

outsinre commented:
@lena-larionova @cloudjumpercat I am doing another round of review based on the latest updates.

outsinre commented on app/_src/gateway/upgrade/lts-upgrade.md:
```diff
@@ -0,0 +1,158 @@
+To make sure your upgrade is successful, carefully review all the steps in this guide. It’s very important to understand all the preparation steps and choose the recommended upgrade path based on your deployment type.
+
+{:.important}
+> **Caution**: The migration pattern described in this document can only happen between two LTS versions, {{site.ee_product_name}} 2.8 LTS and {{site.ee_product_name}} 3.4 LTS. If you apply this document to other release intervals, database modifications may be run in the wrong sequence and leave the database schema in a broken state. To migrate between other versions, see the [general upgrade gu ...
```
```suggestion
> **Caution**: The migration pattern described in this document can only happen between two LTS versions, {{site.ee_product_name}} 2.8 LTS and {{site.ee_product_name}} 3.4 LTS. If you apply this document to other release intervals, database modifications may be run in the wrong sequence and leave the database schema in a broken state. To migrate between other versions, see the [general upgrade guide](/gateway/{{page.kong_version}}/upgrade/).
```

outsinre commented on app/_src/gateway/upgrade/lts-upgrade.md:
```diff
@@ -0,0 +1,158 @@
+Plan a suitable maintenance or downtime window during which you can perform the upgrade.
+During this period, the {{site.base_gateway}} will be temporarily unavailable.
+
+For scenarios where zero downtime is critical, consider the dual-cluster upgrade method, 
```
```suggestion
For scenarios where zero downtime is critical, consider the [dual-cluster upgrade](#dual-cluster-upgrade) method, 
```

outsinre commented on app/_src/gateway/upgrade/dual-cluster.md:
```diff
@@ -0,0 +1,130 @@
+
+5. Deploy a new {{site.base_gateway}} cluster of version Y:
+
+    1. Install a new {{site.base_gateway}} cluster running version Y as instructed in the 
+    [{{site.base_gateway}} Installation Options](/gateway/{{page.kong_version}}/install/).
```
```suggestion
    1. Install a new {{site.base_gateway}} cluster running version Y as instructed in the 
    [{{site.base_gateway}} Installation Options](/gateway/{{page.kong_version}}/install/).
    
      Please provision the new cluster Y with the same-sized resource capacity as that of current cluster X.
```

outsinre commented on app/_src/gateway/upgrade/in-place.md:
```diff
@@ -0,0 +1,92 @@
+5. Stop the {{site.base_gateway}} nodes of the old cluster X but keep the database running. 
+This will create a period of downtime until new nodes are deployed.
+
+6. Install a new cluster running version Y as instructed in the 
+    [{{site.base_gateway}} Installation Options](/gateway/{{page.kong_version}}/install/) and 
+    point it at the existing database for cluster X.
```
```suggestion
6. Install a new cluster running version Y as instructed in the 
    [{{site.base_gateway}} Installation Options](/gateway/{{page.kong_version}}/install/) and 
    point it at the existing database for cluster X.
    
      Please provision the new cluster Y with the same-sized resource capacity as that of current cluster X
```

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,270 @@
+---
+
+Before you start any upgrade, back up your {{site.base_gateway}} data. 
+Kong supports two back up methods for {{site.base_gateway}} entities: database backup and declarative backup.
```
Please ensure to rename "database backup"  to "database-native backup" on this page. Here is an example:

```suggestion
Kong supports two back up methods for {{site.base_gateway}} entities: [database-native backup](#database-native-backup) and [declarative backup](#declarative-backup).
```

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,270 @@
+
+Before you start any upgrade, back up your {{site.base_gateway}} data. 
+Kong supports two back up methods for {{site.base_gateway}} entities: database backup and declarative backup.
+A database backup backs up the entire {{site.base_gateway}} database, while a declarative backup works by managing declarative configuration files.
```
```suggestion
A database-native backup backs up the entire {{site.base_gateway}} database, while a declarative backup works by managing declarative configuration files.
```

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,270 @@
+A database backup backs up the entire {{site.base_gateway}} database, while a declarative backup works by managing declarative configuration files.
+
+We recommend backing up data using both methods, as this offers recovery flexibility:
+* The native database tools are robust and can restore data instantly, compared to the declarative tool. 
```
```suggestion
* The database-native tools are robust and can restore data instantly, compared to the declarative tool. 
```

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,270 @@
+
+    Both tools require the database to be ready for data export and import. To import or export data using these tools, ensure the user and password are initialized, and the database is bootstrapped.
+
+* For **DB-less deployments**, the Kong CLI is the only backup and restore tool available. decK is not supported, as it requires the Admin API to function.
```
```suggestion
* For **DB-less deployments**, neither tools work Please backup the your declarative files manually.
```

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,270 @@
+* **Entities managed by decK**: decK does not manage Enterprise-only entities, like RBAC roles, credentials, keyring, licence, etc. Configure these security related entities separately using Admin API or Kong Manager.
+See the reference for [Entities managed by decK](/deck/latest/reference/entities/) for a full list.
+
+Due to these limitations, we recommend prioritizing the [database native method](#database-backup) in deployments using a database.
```
```suggestion
Due to these limitations, we recommend prioritizing the [database-native method](#database-native-backup) in deployments using a database.
```

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,270 @@
+
+### Database-native backup
+
+When upgrading your {{site.base_gateway}} to a newer version, you have to perform a database migration using the `kong migrations` utility. The `kong migrations` commands are not reversible. We recommend backing up data before any starting any upgrade in case of any migration issues.
```
Add a ref link to `kong migrations` CLI.

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,270 @@
+
+If you are running {{site.base_gateway}} with a database, run a database dump of raw data so that you can recover the database quickly in a database-native way. This is the recommended way to back up {{site.base_gateway}}.
+
+With PostgreSQL, you can dump data in text format, tar format (no compression), or directory format (with compression) using the utility `pg_dump`. For example:
```
```suggestion
With PostgreSQL, you can dump data in *text* format, *tar* format (no compression), or *directory* format (with compression) using the utility `pg_dump`. For example:
```

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,270 @@
+
+### Declarative backup
+
+{% navtabs %}
```
Great style!

lena-larionova commented on app/_includes/md/gateway/hybrid-upgrade.md:
```diff
@@ -0,0 +1,180 @@
+    style CPX stroke-dasharray:3
+    style Admin fill:none,stroke:none,color:#d44324
+    linkStyle 2,3 stroke:#d44324,color:#d44324
+{% endmermaid %}
```
This is difficult to diagram with the software, unfortunately. How about something like this:
![Screenshot 2023-12-07 at 9 05 21 AM](https://github.com/Kong/docs.konghq.com/assets/54370747/1cea8ba7-0403-463f-b9c8-ee8dcdd95422)

lena-larionova commented:
@outsinre I had to get a bit creative to get the diagrams to contain all the elements they need, but I _think_ they're good now. The text-based tool is good for consistency but limited in certain ways.

lena-larionova commented on app/_src/gateway/upgrade/blue-green.md:
```diff
@@ -0,0 +1,136 @@
+    LB --10%--> CPY
+    Admin -."X".- CPY
+    CPX -.-> DB
+    CPY --kong migrations up--> DB
```
unfortunately can't add a strikethrough style, but hopefully the latest changes help.

outsinre commented on app/_src/gateway/upgrade/backup-and-restore.md:
````diff
@@ -0,0 +1,270 @@
+    kong config parse /path/to/kong_backup.yaml
+    ```
+
+2. Restart or reload your {{site.base_gateway}} instance using the backup configuration file:
+
+    ```sh
+    kong restart -c /path/to/kong_backup.yaml
+    ```
+
+    or
+
+    ```sh
+    kong reload -c /path/to/kong_backup.yaml
+    ```
````
This is NOT right! To restore with the declarative backup file, please point [declarative_config](https://docs.konghq.com/gateway/latest/reference/configuration/#declarative_config) in `kong.conf` to `/path/to/kong_backup.yaml`.

```suggestion
2. Restart or reload your {{site.base_gateway}} instance using the backup configuration file:

    ```sh
    export KONG_DECLARATIVE_CONFIG=kong.yml; kong restart -c /path/to/kong_backup.yaml
    # -or-
    export KONG_DECLARATIVE_CONFIG=kong.yml; kong reload -c /path/to/kong_backup.yaml
    ```

    Alternatively, post the declarative backup file to `:8001/config` endpoint.

    ```sh
    curl -sS http://localhost:8001/config?check_hash=1 -F 'config=@kong_backup.yaml' ; echo
    ```
```

outsinre commented:
Unless 1 fundamental fact needs to be updated, LGTM.

Once updated, can be merged. And, wait for customer feedback.

lena-larionova commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,270 @@
+
+## Declarative tools for backup and restore
+
+Kong ships two declarative backup tools: [decK](/deck/) and the [Kong CLI](/gateway/{{page.kong_version}}/reference/cli/), which support managing {{site.base_gateway}} entities in the declarative format. 
```
```suggestion
Kong ships two declarative backup tools: [decK](/deck/) and the [`kong config` CLI](/gateway/{{page.kong_version}}/reference/cli/), which support managing {{site.base_gateway}} entities in the declarative format. 
```

lena-larionova commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,270 @@
+
+* For **DB-less deployments**, no special tools are needed, so there is no declarative tool support. Back up your declarative files manually.
+
+decK is generally more powerful than the Kong CLI. It has more features, invalidates the cache automatically, and fetches entities from the database instead of the LRU cache. Additionally, it overwrites entities instead of patching, so that the database has the exact copy of the config that you provide.
```
```suggestion
decK is generally more powerful than the `kong config` CLI. It has more features, invalidates the cache automatically, and fetches entities from the database instead of the LRU cache. Additionally, it overwrites entities instead of patching, so that the database has the exact copy of the config that you provide.
```

lena-larionova commented on app/_src/gateway/upgrade/backup-and-restore.md:
```diff
@@ -0,0 +1,270 @@
+
+However, decK also has its limitations:
+
+* **Availability**: decK requires {{site.base_gateway}} to be online, while the Kong CLI doesn't. 
```
```suggestion
* **Availability**: decK requires {{site.base_gateway}} to be online, while the `kong config` CLI doesn't. 
```

lena-larionova commented on app/_src/gateway/upgrade/in-place.md:
```diff
@@ -0,0 +1,96 @@
+    linkStyle 0,3,4 stroke:#d44324,color:#d44324
+{% endmermaid %}
+
+> _Figure 1: The diagram shows an in-place upgrade workflow, where the current CP X is directly replaced by a new CP Y.+
+_The database is reused by the new CP Y, and the current CP X is shut down once all nodes are migrated. No Admin API write operations can be performed during the upgrade._
```
```suggestion
> _Figure 1: The diagram shows an in-place upgrade workflow, where the current cluster X is directly replaced by a new cluster Y._
_The database is reused by the new cluster Y, and the current cluster X is shut down once all nodes are migrated. No Admin API write operations can be performed during the upgrade._
```

lena-larionova commented on app/_src/gateway/upgrade/in-place.md:
```diff
@@ -0,0 +1,96 @@
+> _Figure 1: The diagram shows an in-place upgrade workflow, where the current CP X is directly replaced by a new CP Y.+
+_The database is reused by the new CP Y, and the current CP X is shut down once all nodes are migrated. No Admin API write operations can be performed during the upgrade._
+
+There is business downtime as cluster X is stopped during the upgrade process. You must carefully review the upgrade considerations in advance. The prescribed steps below are recommended for you.
```
it's also redundant at this point, a leftover from the very first draft. removing the last sentence.

cloudjumpercat reviewed (approved):
LGTM! Just left a few small suggestions, all just nitpicks though. These guides look really good and I'm looking forward to having these published :)
